cbits/is-valid-utf8.c

```diff
@@ -142,4 +153,544 @@ int bytestring_is_valid_utf8(uint8_t const *const src, size_t const len) {
+      first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[1]));
+      check_block_avx2(inputs[0], prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[1], first_len);
+      prev_first_len = first_len;
+      first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[2]));
+      check_block_avx2(inputs[1], prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[2], first_len);
+      prev_first_len = first_len;
+      first_len =
+          _mm256_shuffle_epi8(first_len_tbl, high_nibbles_of_avx2(inputs[3]));
+      check_block_avx2(inputs[2], prev_first_len, errors, first_range_tbl,
+                       range_min_tbl, range_max_tbl, df_ee_tbl, ef_fe_tbl,
+                       inputs[3], first_len);
+      prev_first_len = first_len;
+    }
+    // Set prev_input based on last block.
+    prev_input = inputs[3];
+    // Advance.
+    ptr += 128;
+  }
+  // Write out the error, check if it's OK.
+  __m256i const combined_errors = _mm256_or_si256(errors[0], errors[1]);
+  if (_mm256_testz_si256(combined_errors, combined_errors) != 1) {
+    return 0;
+  }
+  // 'Roll back' our pointer a little to prepare for a slow search of the rest.
+  uint32_t tokens_blob = _mm256_extract_epi32(prev_input, 7);
+  int8_t const *tokens = (int8_t const *)&tokens_blob;
+  ptrdiff_t lookahead = 0;
+  if (tokens[3] > (int8_t)0xBF) {
+    lookahead = 1;
+  } else if (tokens[2] > (int8_t)0xBF) {
+    lookahead = 2;
+  } else if (tokens[1] > (int8_t)0xBF) {
+    lookahead = 3;
+  }
+  uint8_t const *const small_ptr = ptr - lookahead;
+  size_t const small_len = remaining + lookahead;
+  return is_valid_utf8_fallback(small_ptr, small_len);
+}
+
+#endif
+
+#ifdef __x86_64__
+bool has_sse2() {
+  uint32_t eax = 0, ebx = 0, ecx = 0, edx = 0;
+  __get_cpuid_count(1, 0, &eax, &ebx, &ecx, &edx);
+  // https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits
+  return edx & (1 << 26);
+}
+
+bool has_ssse3() {
+  uint32_t eax = 0, ebx = 0, ecx = 0, edx = 0;
+  __get_cpuid_count(1, 0, &eax, &ebx, &ecx, &edx);
+  // https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits
+  return ecx & (1 << 9);
+}
+
+bool has_avx2() {
+  uint32_t eax = 0, ebx = 0, ecx = 0, edx = 0;
+  __get_cpuid_count(7, 0, &eax, &ebx, &ecx, &edx);
+  // https://en.wikipedia.org/wiki/CPUID#EAX=7,_ECX=0:_Extended_Features
+  return ebx & (1 << 5);
+}
+#endif
+
+typedef int (*is_valid_utf8_t) (uint8_t const *const, size_t const);
+
+int bytestring_is_valid_utf8(uint8_t const *const src, size_t const len) {
+  if (len == 0) {
+    return 1;
+  }
+#ifdef __x86_64__
+  static _Atomic is_valid_utf8_t s_impl = (is_valid_utf8_t)NULL;
+  is_valid_utf8_t impl = atomic_load_explicit(&s_impl, memory_order_relaxed);
```

Comment:
I've tried this kind of approach before, and it caused _massive_ performance degradation. Did you check the assembly this outputs?

Reply:
I don't know how to compare assembly, but I run `text` benchmarks before and after, no measurable difference.

Reply:
OK, in that case, I think it's fine.

Reply:
Thanks for review!